crate/operator/cratedb.py

```diff
@@ -125,86 +124,43 @@ async def get_healthiness(cursor: Cursor) -> int:
-                    logger.info("Waiting for cluster to get healthy again ...")
-                    try:
-                        num_nodes = await get_number_of_nodes(cursor)
-                        healthiness = await get_healthiness(cursor)
-                        if num_nodes == expected_nodes and healthiness in {1, None}:
-                            healthy_count += 1
-                            logger.info(
-                                "Cluster has expected nodes and is healthy (%d of 3)",
-                                healthy_count,
-                            )
-                        else:
-                            logger.info(
-                                "Cluster has %d of %d nodes and is in %s state.",
-                                num_nodes,
-                                expected_nodes,
-                                HEALTHINESS.get(healthiness, "UNKNOWN"),
-                            )
-                            healthy_count = 0
-                    except ProgrammingError as e:
-                        logger.warning("Failed to run health check query", exc_info=e)
-                        # We decided to endlessly retry. The reason for that
-                        # is, that an issue with cluster health could be solved
-                        # by intervening on the cluster directly, e.g. by
-                        # adjusting the number of replicas, and in this case
-                        # the operator should not fail but continue.
-                        healthy_count = 0
-                        await asyncio.sleep(BACKOFF_TIME)
-                    else:
-                        if healthy_count >= 3:
-                            break
-                        else:
-                            # Wait some time before we retry.
-                            await asyncio.sleep(BACKOFF_TIME / 2.0)
-        except Exception as e:
-            # Sometimes the client is not able to connect to a CrateDB cluster,
-            # e.g. when a load balancer is timing out, or when a load balancer
-            # routes to an non-existent node. In this case we retry 5 times and
-            # only fail then.
-            if exception_count > 5:
-                raise e
-            else:
-                healthy_count = 0
-                logger.warning("Failed to connect to cluster", exc_info=e)
-            exception_count += 1
-            await asyncio.sleep(BACKOFF_TIME)
+    # We need to establish a new connection because the peer of a
+    # previous connection could have been shut down. And by
+    # re-establishing a connection for _each_ polling we can assert
+    # that the connection is open
+    async with connection_factory() as conn:
+        async with conn.cursor() as cursor:
+            logger.info("Checking if cluster is healthy ...")
+            try:
+                num_nodes = await get_number_of_nodes(cursor)
+                healthiness = await get_healthiness(cursor)
```

Comment:
Good point. I think yes.

Reply:
Actually, not sure. Since they both use the same `cursor` object, wouldn't that potentially cause conflicts? Since both functions run `cursor.execute()` and `cursor.fetchone()`, couldn't the `fetchone()` in `get_healthiness()` return the response for the query in `get_number_of_nodes()`, if that's `execute()` was run later?

```
 get_number_of_nodes() | get_healthiness()   | What would cursor.fetchone() return
--------------------------------------------------------------------------------------------------
 cursor.execute(...)   |                     |
                       |                     | Result for number of nodes
                       | cursor.execute(...) |
                       |                     | Result for healthiness
 cursor.fetchone()     |                     | Returns the healthiness instead of number of nodes
                       | cursor.fetchone()   | Returns the healthiness instead of number of nodes


 get_number_of_nodes() | get_healthiness()   | What would cursor.fetchone() return
--------------------------------------------------------------------------------------------------
 cursor.execute(...)   |                     |
                       |                     | Result for number of nodes
 cursor.fetchone()     |                     | Returns the number of nodes
                       | cursor.execute(...) |
                       |                     | Result for healthiness
                       | cursor.fetchone()   | Returns the healthiness instead of number of nodes


 get_number_of_nodes() | get_healthiness()   | What would cursor.fetchone() return
--------------------------------------------------------------------------------------------------
                       | cursor.execute(...) |
                       |                     | Result for healthiness
 cursor.execute(...)   |                     |
                       |                     | Result for number of nodes
 cursor.fetchone()     |                     | Returns the number of nodes
                       | cursor.fetchone()   | Returns the number of nodes instead of healthiness
```

Reply:
Right, you'd need separate cursor `objects`.